### DISCUSSION TITLE
Multiple Drumkits in One Project?

### DISCUSSION THREAD

**Discussion by charlescochran** [hydrogen-music/hydrogen#1087] [Q&A] [answered]
Hi,

I recently discovered Hydrogen and (with an increasing level of excitement) have been learning about it and its capabilities. In the process of this, I found myself naturally wanting to load drumkits on a per pattern basis. I understand that you can load multiple instruments from multiple saved drumkits simultaneously, but _all_ of the instruments will then show up within all of the patterns, which could be cumbersome. My intuition and a search through the manual came up dry, so I'm wondering, has this feature not been implemented, or what am I missing? (If it hasn't, I have a feeling it might be useful. Can't back that up with experience yet, of course.)

Thanks!

EDIT: For those interested, I laid out my ideas for this feature in a comment over in #679.

**Comment by oddtime** [comment, marked as answer]
Hi,
I don't think that you can load a whole drumkit at once without replacing the previously loaded instruments.
This could be a feature request.
Then both drumkits would be visible and available in every pattern, unless we don't implement a custom instrument visibility (or order) for each pattern, which would be another feat. request.

**Reply by charlescochran** [reply]
Gotcha. Do you think this warrants adding a feature request for? It might not be as important as it seemed to me, given that it hasn't been implemented yet.

**Reply by oddtime** [reply]
I would search if there are open issues about it. I read something similar...

**Reply by charlescochran** [reply]
#679 Yep... Thanks.

**Comment by trebmuh** [comment]
@charlescochran : depending of your use case, you might be able to launch 2 different instances of Hydrogen, each one loaded with pattern/song and different drumkits, and play them together using the jack-transport functionality.

**Reply by charlescochran** [reply]
Ah, that's a neat idea. I'll definitely give it a try if I find it necessary. Thanks!

**Comment by arnwas** [comment]
wouldn't it be easier if hydrogen would internally concatenate different drumkits, but showing them separately on diferent patterns along users' choice?

**Reply by theGreatWhiteShark** [reply]
> wouldn't it be easier if hydrogen would internally concatenate different drumkits, but showing them separately on diferent patterns along users' choice?

Concatenating drumkits or holding a couple of them per song is no problem at all - at least internally.

What makes trouble is the change in UX. A simple drop down menu, like the one for the components, would give a horrible UX and would make the whole application more complex and harder to use. Instead, larger parts of both pattern and song editor have to be redesign and changed. And this likely requires 1-2 months (at least for me).

As I do not clearly see the intended use case yet: Is it about different patterns (played back at the same time or used in pattern mode) holding different kits or rather about different parts of a song being governed by individual kits? In case of the latter are more clear (and easy) approach would be to allow switching of drumkits in a programmable way.

**Comment by charlescochran** [comment]
> Is it about different patterns (played back at the same time or used in pattern mode) holding different kits or rather about different parts of a song being governed by individual kits?

Perhaps both, but especially the former. Programmatic switching might be neat, but I think the biggest desire would be to have loading a drumkit (optionally) apply to the current pattern rather than the current song. This would address both.

**Reply by theGreatWhiteShark** [reply]
> I think the biggest desire would be to have loading a drumkit (optionally) apply to the current pattern rather than the current song

Hmm.. TBH I do not like the idea at all. Opening a song containing patterns with various drumkits assigned to them would make Hydrogen load all those kits at once. (Otherwise we would get audio glitches, delayes etc. when switching from one kit to another). For the older kits this is not a big deal. "TR808EmulationKit" for example is just 244KB. But "BigMono" is 213MB and there are SFZ kits out there which are easily larger than 1GB. Attempting to load a song containing a couple of patterns using such kits - or maybe just importing a pattern from the internet that is using one of them - could make Hydrogen allocate so much RAM it freezes your computer. Worst case, this then will happen every single time you start up the application (and load the last loaded song). Allowing to load an arbitrary number of kits at once (as we can have an arbitrary number of patterns) would be a massive memory leak almost impossible to handle properly. With just one kit the user only has to assure her computer can handle the largest kit

**Reply by charlescochran** [reply]
I see. Maybe https://github.com/hydrogen-music/hydrogen/issues/679#issuecomment-757505107 could be a way to go, then.

**Comment by osirisgothra** [comment]
Just chiming in to say yes this is a feature we would want. Basically (in my perspective) you wouldn't need to do more than just automate the process of dragging each drum in the desired sound library kit to the bottom of the loaded kit which already works. This is in fact simple enough that I have a macro that does this for me, but that solution is not sound because if the window is not set up exactly the same each time, the mouse would click the wrong place. 

So yeah want this, even a keyboard shortcut to navigate and append each drum one by one would be a step up from what we currently have to do because its a ..."major drag" lol.  I was thinking of just making a patch myself but I have many projects that take priority that are ongoing and that would slow all those down :3 Now Im off to see if there is any other issues/requests about this...

**Reply by theGreatWhiteShark** [reply]
Hey @osirisgothra,

So, you suggest that - in addition to drag&drop individual instruments into the song's drumkit - the user should be able to import _all_ instruments by drag&drop the corresponding drumkit node? Do I get your idea right? This definitely would be a nice addition.

**Reply by theGreatWhiteShark** [reply]
@osirisgothra in version 2.0 multiple instruments can be selected and imported at once
